File: 606. Construct String from Binary Tree/606.go
package leetcode

import(
	"strings"
	"strconv"
)

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func tree2str(t *TreeNode) string {
	res := make([]string, 0)
	if t == nil{
		return ""
	}
	res = append(res, strconv.Itoa(t.Val))
	if t.Left != nil || t.Right != nil{
		tree(t.Left, &res)
		if t.Right != nil{
			tree(t.Right, &res)
		}
	}
	return strings.Join(res, "")
}

func tree(t *TreeNode, res *[]string){
	if t == nil{
		*res = append(*res, "()")
	}else{
		*res = append(*res, "(")
		*res = append(*res, strconv.Itoa(t.Val))
		if t.Left != nil || t.Right != nil{
			tree(t.Left, res)
			if t.Right != nil{
				tree(t.Right, res)
			}
		}
		*res = append(*res, ")")
	}
}

func tree2str1(t *TreeNode) string {
    
    if t == nil { return "" }
    s := strconv.Itoa(t.Val)
    l := tree2str(t.Left)
    r := tree2str(t.Right)
    
    if t.Left == nil && t.Right == nil {
        return s
    }
    if t.Right == nil {
        return s + "(" + l + ")"
    }
    
    return s + "(" + l + ")" + "(" + r + ")"
    
}

func tree2str2(t *TreeNode) string {
	if t == nil {
		return ""
	}
	if t.Left == nil && t.Right == nil {
		return strconv.Itoa(t.Val)
	}
	if t.Left != nil && t.Right == nil {
		return strconv.Itoa(t.Val) + "(" + tree2str(t.Left) + ")"
	}

	if t.Left == nil && t.Right != nil {
		return strconv.Itoa(t.Val) + "()" + "(" + tree2str(t.Right) + ")"
	}

	return strconv.Itoa(t.Val) + "(" + tree2str(t.Left) + ")(" + tree2str(t.Right) + ")"
}

File: 606. Construct String from Binary Tree/_606.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def tree2str(self, t: TreeNode) -> str:
        res = list()
        if not t:
            return ""
        res.append(str(t.val))
        if t.left or t.right:
            self.tree(t.left, res)
            if t.right:
                self.tree(t.right,res)
        return ''.join(res)


    def tree(self, t: TreeNode, res: list):
        if not t:  
            res.append('()')
        else:
            res.append('(')
            res.append(str(t.val))
            if t.right or t.left:
                self.tree(t.left, res)
                if t.right:
                    self.tree(t.right,res)
            res.append(')')

    def tree2str1(self, t):
        if not t: return ''
        left = '({})'.format(self.tree2str(t.left)) if (t.left or t.right) else ''
        right = '({})'.format(self.tree2str(t.right)) if t.right else ''
        return '{}{}{}'.format(t.val, left, right)

    def tree2str2(self, t):
        if t is None:
                return ""
        ans = str(t.val) 
        if t.left:
            ans+="(" + self.tree2str(t.left) + ")"
        if not t.left and t.right:
            ans+="()"
        if t.right:
            ans+="(" + self.tree2str(t.right) + ")"
        return ans